core/src/test/java/org/elasticsearch/threadpool/ScheduleWithFixedDelayTests.java

```diff
@@ -0,0 +1,287 @@
+            threadList.add(new Thread() {
+                @Override
+                public void run() {
+                    latch.countDown();
+                    try {
+                        latch.await();
+                        cancellable.cancel();
+                    } catch (InterruptedException e) {
+                        Thread.currentThread().interrupt();
+                    }
+                }
+            });
+        }
+
+        for (Thread thread : threadList) {
+            thread.start();
+        }
+        latch.countDown();
+        for (Thread thread : threadList) {
+            thread.join();
+        }
+
+        assertTrue(cancellable.isCancelled());
+    }
+
+    public void testBlockingCallOnSchedulerThreadFails() throws Exception {
+        final AtomicReference<Throwable> throwableRef = new AtomicReference<>();
+        final BaseFuture<Object> future = new BaseFuture<Object>() {};
+        final Runnable runnable = () -> {
+            try {
+                if (randomBoolean()) {
+                    future.get();
+                } else {
+                    future.get(1L, TimeUnit.SECONDS);
+                }
+            } catch (Throwable t) {
+                throwableRef.set(t);
+            }
+        };
+
+        Cancellable cancellable = threadPool.scheduleWithFixedDelay(runnable, TimeValue.timeValueMillis(100L), Names.SAME);
+        boolean failed = awaitBusy(() -> throwableRef.get() != null, 1L, TimeUnit.SECONDS);
+        assertTrue(failed);
+        assertThat(throwableRef.get(), instanceOf(AssertionError.class));
+        assertThat(throwableRef.get().getMessage(), containsString("Blocking"));
+        assertFalse(cancellable.isCancelled());
+    }
+
+    public void testBlockingCallOnNonSchedulerThreadAllowed() throws Exception {
+        class TestFuture extends BaseFuture<Object> {
+            public boolean futureDone(Object value) {
+                return set(value);
+            }
+        }
+
+        final AtomicReference<Throwable> throwableRef = new AtomicReference<>();
+        final AtomicReference<Object> objectRef = new AtomicReference<>();
+        final TestFuture future = new TestFuture();
+        final boolean rethrow = randomBoolean();
+
+        final Runnable runnable = () -> {
+            try {
+                if (randomBoolean()) {
+                    objectRef.set(future.get());
+                } else {
+                    objectRef.set(future.get(1, TimeUnit.MINUTES));
+                }
+            } catch (Throwable t) {
+                throwableRef.set(t);
+                if (rethrow) {
+                    throw new RuntimeException(t);
+                }
+            }
+        };
+
+        final Cancellable cancellable = threadPool.scheduleWithFixedDelay(runnable, TimeValue.timeValueMillis(20L), Names.GENERIC);
+        boolean failed = awaitBusy(() -> throwableRef.get() != null, 200, TimeUnit.MILLISECONDS);
```

Comment:
I'm not a fan of this. we effectively always wait. Can we just rely on the `assertNull(throwableRef.get())` at the end of the test? it may have some false positives but if we get something wrong, we'll know soon enough